### DISCUSSION TITLE
Customize content and looks of email templates

### DISCUSSION THREAD

**Discussion by saibotma** [appwrite/appwrite#3576] [Ideas]
## 🚀 Feature

As an admin I want to be able to customize the content and looks of email templates dynamically in the admin interface.

## Have you spent some time to check if this issue has been raised before?

yes

### Have you read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/master/CODE_OF_CONDUCT.md)?

yes

## Pitch

Emails can be branded with the company design.

**Comment by TorstenDittmann** [comment]
I already set up a good foundation for this issue on the template side [here](https://github.com/appwrite/appwrite/pull/402).

Files that can be customized for this:
https://github.com/appwrite/appwrite/blob/master/app/config/locales/templates/_cta.tpl
https://github.com/appwrite/appwrite/blob/master/app/config/locales/templates/_base.tpl

Also there are some things in place already for some kind of customization like

```
->setParam('{{bg-body}}', '#f6f6f6')
->setParam('{{bg-content}}', '#ffffff')
->setParam('{{bg-cta}}', '#3498db')
->setParam('{{bg-cta-hover}}', '#34495e')
->setParam('{{text-content}}', '#000000')
->setParam('{{text-cta}}', '#ffffff')
```

Things left to add for this:
- Pages in the dashboard
- A way to save custom HTML/colors in the database
- Use a custom template for the mails if set

I doubt we need any endpoints.

Do we want to add only color branding or complete HTML customization?

**Comment by eldadfux** [comment]
Umm.. maybe we can add the colors as project settings (alongside the project logo), which will be consumed from the default template and also add an option to completely overwrite the templates HTML code and contents?

We also need to think about multi-locale content, should we allow content overwriting for each locale separately?

**Comment by TorstenDittmann** [comment]
> Umm.. maybe we can add the colors as project settings (alongside the project logo), which will be consumed from the default template and also add an option to completely overwrite the templates HTML code and contents?

Had the same thought. Maybe we also add a {logo} variable. CSS in these templates is a pain - so I doubt it is a good idea for many people. Would still love to have it.

> We also need to think about multi-locale content, should we allow content overwriting for each locale separately?

Mhh, I only thought of global changes. Also can't think of a use case for this 🤔

**Comment by eldadfux** [comment]
> Mhh, I only thought of global changes. Also can't think of a use case for this 🤔

For example, if an app wants to support 3 main languages, and the developer wants to have customised email messages. Writing the down makes me think this might be a rare edge case or at least too early to think about.

**Comment by saibotma** [comment]
@TorstenDittmann the optimal solution would be if the user could just upload a template created with the templating language appwrite uses. This way appwrite would not need to have any editor for the templates.

But being able to change colors, font and content would be enough for the beginning I believe.

**Comment by TorstenDittmann** [comment]
That means we would need to overthink, on how we store translations for these mails. Or translations in general.

I would also love to introduce something like [MJML](https://mjml.io/), which would bring the e-mails send by Appwrite to the next level. This would also make custom e-mail templates very accessible - since in todays world we have to use some sketchy CSS attributes and tables in HTML to make emails look good.

MJML would introduce a well documented templating syntax and we just provide the placeholders for the content.

@eldadfux there is the scary part, we would introduce a dependency, but that doesnt mean that MJML is mandatory for the user. Since you could just use custom HTML/CSS - run it throught the interpreter and nothing changed.

**Comment by brandonroberts** [comment]
@saibotma 👋🏿  following up on this issue. Is this still relevant for you?

**Comment by JulianBissekkou** [comment]
Just want to mention that this is a feature that is relevant for us as well 👍🏽

**Comment by saibotma** [comment]
> @saibotma 👋🏿  following up on this issue. Is this still relevant for you?

Actually not really at the moment.

**Comment by saibotma** [comment]
Yeah @brandonroberts, @JulianBissekkou should now be the contact partner, as I am working on a different project now. Totally forgot this.

**Comment by Aragur** [comment]
> @saibotma 👋🏿 following up on this issue. Is this still relevant for you?

It's also relevant for me.  
Currently I'm using my own service to send customized emails using a cloud function that is triggered by all the events.  
Not the cleanest solution and realy buggy atm.  
So it would be verry nice to have a feature like this.

**Comment by brandonroberts** [comment]
@Aragur using cloud functions is buggy or something else?

**Comment by Aragur** [comment]
@brandonroberts sorry for the late answer.
The problem was (or maybe still is) that not all actions, which should send an email,  had a trigger to call the function.

Also I think that if you want to send a e-mail for user verification you shouldn't depend on a cloud function which may fail.

I think a basic implementation to modify or to update the look of the email would be really nice. (Maybe like [Gitea](https://docs.gitea.io/en-us/mail-templates/)).
If I'm developing an app I at least want to send my user greetings from my app if he signs up and a verification email was sent.

**Comment by tessamero** [comment]
Hey all, just an update, that this is a plan for the (near) future, but not officially on the roadmap for this year as we have a lot of other major priorities to accomplish that the community wants to see. 

This would also require a lot more discussion. I'd like to see this moved to Discussions :) We can collect any feedback over time to help us with this feature and it'll help get it on the roadmap sooner

**Comment by cliftonlabrum** [comment]
If we can't have custom email templates, what about providing APIs to generate the content of the email and then let us do whatever we want with it?

For example, in the case of magic link authentication, Supabase lets you generate the link itself in the SDK, then you can take it and send it off to Sendgrid (or whatever) to use inside your own custom template. 😊

In the end, I don't think you need to spend a ton of time making an email template. Just give us a place to paste in our HTML and then use some kind of replacement identified that your script uses to fill in the link:
```
<p>My template here. <a href="{{magicLinkURL}}">Link</a></p>
```

**Comment by eldadfux** [comment]
I'm happy to share this is now at work, and is expected to be released in the next Appwrite version.

**Comment by pratikbaid3** [comment]
Hi @eldadfux . Is this now available? If so, is there a example implementation for it?
